File: Go/Linked List/19_Remove Nth Node From End of List.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func removeNthFromEnd(head *ListNode, n int) *ListNode {
	if head == nil || (head.Next == nil && n == 1) {
		return nil
	}
	p2 := &ListNode{Next: head}
	count := n //2
	for i := 0; i < count; i++ {
		p2 = p2.Next
	}
	p1 := &ListNode{Next: head}
	for p2.Next != nil {
		p1, p2 = p1.Next, p2.Next
	}
	p1.Next = p1.Next.Next
	return head
}

File: C++/Linked List/19_Remove_Nth Node_From_End_of_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    ListNode *removeNthFromEnd(ListNode *head, int n)
    {
        // find nth Node from the end
        ListNode *dummy = new ListNode(0, head);
        ListNode *first = dummy, *second = dummy;
        // traverse first pointer to n+1 so that distance between first and second is n
        // nodes apart
        for (int i = 1; i <= n + 1; i++)
        {
            first = first->next;
        }
        // Move first to the end, maintaining the gap
        while (first != NULL)
        {
            first = first->next;
            second = second->next;
        }
        // remove nth node(second->next)
        second->next = second->next->next;
        return dummy->next;
    }
};
